File: pnix-hy/pnix-hy/pnix_hy/incremental.py
"""pnix_hy.incremental -- definition-granular content addressing + realisation early cutoff
(proposal 0023; Unison identity model + Nix CA resolved derivations).

R1: each top-level `let` definition gets a DEPENDENCY-SUBSTITUTED content hash -- references to
sibling definitions are replaced by those definitions' hashes before hashing, so a definition's
identity depends on its meaning, not on sibling NAMES (alpha-renaming a dependency does not
invalidate anything: names are metadata). `incremental_eval` caches each pure, data-valued
definition by that hash and recomputes ONLY definitions whose hash changed (plus dependents).

R3: `realisation_record` is the Nix-CA analogue -- a store mapping `ir_sha256` (the drv) to the
produced `value_hash` (the out) with a witness; a known ir hash short-circuits evaluation
entirely (early cutoff).

Soundness first: anything unsupported (non-let programs, cyclic definitions, impure source,
non-data definition values) falls back to a full evaluation -- a cache miss is safe, a wrong
hit is not.
"""

from __future__ import annotations

import hashlib
import json
from typing import Any

from . import pnix_runtime as rt

_DEF_CACHE: dict[str, Any] = {}
_REALISATIONS: dict[str, dict[str, Any]] = {}


def clear_incremental_cache() -> None:
    _DEF_CACHE.clear()
    _REALISATIONS.clear()


def _collect_refs(node: Any, names: frozenset[str], out: set[str]) -> None:
    if isinstance(node, dict):
        if node.get("tag") == "var" and node.get("name") in names:
            out.add(node["name"])
        for v in node.values():
            _collect_refs(v, names, out)
    elif isinstance(node, list):
        for v in node:
            _collect_refs(v, names, out)


def _substitute_refs(node: Any, hashes: dict[str, str]) -> Any:
    """Replace references to sibling definitions by their HASHES (Unison: dependencies are
    replaced by their hashes before hashing, so names become metadata)."""
    if isinstance(node, dict):
        if node.get("tag") == "var" and node.get("name") in hashes:
            return {"tag": "__ref__", "hash": hashes[node["name"]]}
        return {k: _substitute_refs(v, hashes) for k, v in node.items()}
    if isinstance(node, list):
        return [_substitute_refs(v, hashes) for v in node]
    return node


def definition_hashes(source: str) -> dict[str, Any]:
    """Dependency-substituted content hash per top-level let definition. `supported: False`
    for non-let programs, multi-segment paths, or cyclic definition graphs."""
    try:
        ast = rt.parse(source)
    except Exception as exc:  # noqa: BLE001
        return {"supported": False, "reason": f"parse: {exc}"}
    if not isinstance(ast, dict) or ast.get("tag") != "let":
        return {"supported": False, "reason": "not a top-level let"}
    bindings = ast.get("bindings") or []
    if any(len(b.get("path", [])) != 1 or not isinstance(b["path"][0], str) for b in bindings):
        return {"supported": False, "reason": "multi-segment binding path"}
    names = frozenset(b["path"][0] for b in bindings)
    values = {b["path"][0]: b["value"] for b in bindings}
    deps: dict[str, set[str]] = {}
    for name, value in values.items():
        refs: set[str] = set()
        _collect_refs(value, names, refs)
        deps[name] = refs - {name}  # a self-reference would be recursive -> cycle below
    hashes: dict[str, str] = {}
    order: list[str] = []
    remaining = set(names)
    while remaining:
        ready = sorted(n for n in remaining if deps[n] <= set(hashes))
        if not ready:
            return {"supported": False, "reason": f"cyclic definitions: {sorted(remaining)}"}
        for name in ready:
            substituted = _substitute_refs(values[name], hashes)
            hashes[name] = rt.ast_hash(substituted)
            order.append(name)
            remaining.discard(name)
    return {"supported": True, "hashes": hashes, "order": order, "deps": {k: sorted(v) for k, v in deps.items()},
            "values": values, "body": ast.get("body")}


def _is_data(value: Any) -> bool:
    try:
        json.dumps(value)
        return True
    except (TypeError, ValueError):
        return False


def incremental_eval(source: str) -> dict[str, Any]:
    """Evaluate a top-level let with a per-definition content-addressed cache: unchanged
    definitions (by dependency-substituted hash) are reused; only changed definitions and
    their dependents are recomputed. Unsupported shapes fall back to full evaluation."""
    info = definition_hashes(source)
    from . import pnix_mirror as pm  # noqa: PLC0415 - lazy: avoid import cycle

    if not info.get("supported") or not pm.static_purity_check(source).get("pure", False):
        value = rt.stable_data(rt.eval_source(source))
        return {"schema": "pnix-hy.incremental-eval.v0", "value": value, "supported": False,
                "reason": info.get("reason", "impure source"), "hits": 0, "misses": 0}
    hits = misses = 0
    env: dict[str, Any] = {}
    for name in info["order"]:
        h = info["hashes"][name]
        if h in _DEF_CACHE:
            env[name] = _DEF_CACHE[h]
            hits += 1
            continue
        ctx = rt.runtime_context(None)
        ctx["env"] = dict(env)  # topological order: every dependency is already materialized
        val = rt.realize_value(rt.force_value(
            rt.eval_source_raw(rt.emit_source(info["values"][name]), ctx, realize=False)))
        if not _is_data(val):  # a function-valued definition cannot be cached as data -> full eval
            value = rt.stable_data(rt.eval_source(source))
            return {"schema": "pnix-hy.incremental-eval.v0", "value": value, "supported": False,
                    "reason": f"definition {name!r} is not data", "hits": hits, "misses": misses}
        _DEF_CACHE[h] = val
        env[name] = val
        misses += 1
    ctx = rt.runtime_context(None)
    ctx["env"] = env
    value = rt.realize_value(rt.force_value(
        rt.eval_source_raw(rt.emit_source(info["body"]), ctx, realize=False)))
    return {"schema": "pnix-hy.incremental-eval.v0", "value": value, "supported": True,
            "hits": hits, "misses": misses, "definition_hashes": info["hashes"]}


def realisation_record(source: str) -> dict[str, Any]:
    """R3 (Nix CA): the `ir_sha256 -> value_hash` realisation store. A known ir hash proves the
    result without evaluating (early cutoff); a miss evaluates once, records the realisation,
    and stamps a witness."""
    from . import gate  # noqa: PLC0415
    from . import ir as ir_mod  # noqa: PLC0415

    drv = ir_mod.ir_of(source)["ir_sha256"]
    if drv in _REALISATIONS:
        return {**_REALISATIONS[drv], "early_cutoff": True}
    value = rt.stable_data(rt.eval_source(source))
    value_hash = hashlib.sha256(
        json.dumps(value, sort_keys=True, separators=(",", ":"), default=repr).encode("utf-8")
    ).hexdigest()
    witness = gate.make_witness("realisation", {"drv": drv, "out": value_hash})
    rec = {"schema": "pnix-hy.realisation.v0", "ir_sha256": drv, "value_hash": value_hash,
           "value": value, "witness_id": witness["witness_id"], "early_cutoff": False}
    _REALISATIONS[drv] = rec
    return rec


def incremental_eval_report() -> dict[str, Any]:
    """Self-check (proposal 0023): full hit on repeat, partial recompute on a one-definition
    change, alpha-rename immunity (names are metadata), realisation early cutoff, and value
    agreement with ground-truth evaluation throughout."""
    try:
        clear_incremental_cache()
        src = "let big = 1000 * 1000; dep = big + 1; other = 7; in dep + other"
        truth = rt.stable_data(rt.eval_source(src))
        e1 = incremental_eval(src)
        e2 = incremental_eval(src)
        cold_warm = (e1["value"] == truth == e2["value"]
                     and e1["misses"] == 3 and e1["hits"] == 0
                     and e2["hits"] == 3 and e2["misses"] == 0)

        # change ONE independent definition -> exactly one recompute
        src_b = "let big = 1000 * 1000; dep = big + 1; other = 8; in dep + other"
        e3 = incremental_eval(src_b)
        partial = e3["value"] == rt.stable_data(rt.eval_source(src_b)) \
            and e3["misses"] == 1 and e3["hits"] == 2

        # alpha-rename a dependency -> ALL definitions still hit (names are metadata)
        src_r = "let huge = 1000 * 1000; dep = huge + 1; other = 8; in dep + other"
        e4 = incremental_eval(src_r)
        alpha = e4["value"] == e3["value"] and e4["hits"] == 3 and e4["misses"] == 0

        # unsupported/impure shapes fall back soundly
        fb = incremental_eval("(x: x + 1) 41")
        fallback_ok = fb["supported"] is False and fb["value"] == 42

        # R3: realisation early cutoff
        r1 = realisation_record("let a = 6; in a * 7")
        r2 = realisation_record("let a = 6; in a * 7")
        real_ok = (r1["early_cutoff"] is False and r2["early_cutoff"] is True
                   and r1["value_hash"] == r2["value_hash"] and r1["value"] == 42
                   and bool(r1["witness_id"]))

        ready = bool(cold_warm and partial and alpha and fallback_ok and real_ok)
        return {"schema": "pnix-hy.incremental-eval.report.v0", "ready": ready, "available": True,
                "cold_then_warm": cold_warm, "partial_recompute": partial,
                "alpha_rename_immune": alpha, "sound_fallback": fallback_ok,
                "realisation_early_cutoff": real_ok}
    except Exception as exc:  # noqa: BLE001
        return {"schema": "pnix-hy.incremental-eval.report.v0", "ready": False,
                "available": False, "error": f"{type(exc).__name__}: {exc}"}


__all__ = ["definition_hashes", "incremental_eval", "realisation_record",
           "clear_incremental_cache", "incremental_eval_report"]

File: pnix-rs/pnix-rs/src/incremental.rs
//! Definition-granular content addressing + realisation early cutoff (P9).
//!
//! Unison identity model: each top-level `let` definition gets a
//! DEPENDENCY-SUBSTITUTED content hash — references to sibling definitions are
//! replaced by those definitions' hashes before hashing, so a definition's
//! identity depends on its meaning, not on sibling NAMES (alpha-renaming a
//! dependency does not invalidate anything: names are metadata). Mutual
//! recursion hashes as an SCC group: members joined in sorted-name order into
//! one group text (v0 honesty: names *inside* a cycle are part of the group
//! text, so renaming within a cycle changes the group hash — documented
//! boundary, not hidden).
//!
//! Nix-CA realisation model: `realisation record` maps `ir_sha256` (the drv)
//! to the produced `value_sha256` (the out) with a witness hash; a known ir
//! hash short-circuits evaluation entirely (early cutoff). The store is a
//! cache (default `work/realisations.tsv`), not a proof receipt.

use crate::gate;
use crate::interop;
use crate::ir;
use crate::px;
use crate::sha256::sha256_hex;

pub const INCREMENTAL_SCHEMA: &str = "pnix-rs.incremental.v0";
pub const DEFAULT_STORE: &str = "work/realisations.tsv";

/// Dependency-substituted hashes for the top-level let definitions.
/// Returns (name, hash) pairs in source order.
pub fn definition_hashes(source: &str) -> Result<Vec<(String, String)>, String> {
    let ast = px::px_parse(source)?;
    let (bindings, _body) = match &ast {
        px::PxExpr::LetIn { bindings, body } => (bindings.clone(), body),
        _ => return Err(String::from("incremental: top-level expression is not a let")),
    };
    let n = bindings.len();
    let names: Vec<String> = bindings.iter().map(|(name, _)| name.clone()).collect();
    for i in 0..n {
        for j in (i + 1)..n {
            if names[i] == names[j] {
                return Err(format!(
                    "incremental: duplicate definition name {} (shadowing lets are not content-addressable)",
                    names[i]
                ));
            }
        }
    }

    // Sibling reference graph.
    let mut refs: Vec<Vec<usize>> = Vec::new();
    for (_name, expr) in &bindings {
        let mut free = Vec::new();
        let mut bound = Vec::new();
        crate::specialize::px_free_vars(expr, &mut free, &mut bound);
        let mut row = Vec::new();
        for (j, sib) in names.iter().enumerate() {
            if free.iter().any(|f| f == sib) {
                row.push(j);
            }
        }
        refs.push(row);
    }

    // Transitive reachability (small n; O(n^3) is fine).
    let mut reach: Vec<Vec<bool>> = vec![vec![false; n]; n];
    for (i, row) in refs.iter().enumerate() {
        for &j in row {
            reach[i][j] = true;
        }
    }
    let mut changed = true;
    while changed {
        changed = false;
        for i in 0..n {
            for j in 0..n {
                if reach[i][j] {
                    for k in 0..n {
                        if reach[j][k] && !reach[i][k] {
                            reach[i][k] = true;
                            changed = true;
                        }
                    }
                }
            }
        }
    }
    // SCC id per node: mutual reachability (or self for non-recursive nodes).
    let mut scc_id: Vec<usize> = (0..n).collect();
    for i in 0..n {
        for j in 0..n {
            if i != j && reach[i][j] && reach[j][i] && scc_id[j] > scc_id[i] {
                scc_id[j] = scc_id[i];
            }
        }
    }

    let canonical =
        |expr: &px::PxExpr, hashes: &Vec<Option<String>>| -> String {
            let substituted = substitute_sibling_refs(expr, &names, hashes);
            px::px_emit(&px::px_normalize(&substituted))
        };

    let mut hashes: Vec<Option<String>> = vec![None; n];
    // Iterate until every definition is hashed (dependencies-first converges
    // because substitution only needs hashes of OTHER SCCs).
    let mut progress = true;
    while progress {
        progress = false;
        for i in 0..n {
            if hashes[i].is_some() {
                continue;
            }
            let group: Vec<usize> = (0..n).filter(|&j| scc_id[j] == scc_id[i]).collect();
            // Ready when every external dependency of the group is hashed.
            let ready = group.iter().all(|&m| {
                refs[m]
                    .iter()
                    .all(|&d| scc_id[d] == scc_id[i] || hashes[d].is_some())
            });
            if !ready {
                continue;
            }
            let self_recursive = group.len() > 1
                || refs[i].iter().any(|&d| d == i);
            if !self_recursive && group.len() == 1 {
                let text = canonical(&bindings[i].1, &hashes);
                hashes[i] = Some(sha256_hex(format!("def:{}", text).as_bytes()));
            } else {
                let mut members: Vec<usize> = group.clone();
                members.sort_by(|a, b| names[*a].cmp(&names[*b]));
                let mut group_text = String::from("scc:");
                for &m in &members {
                    let text = canonical(&bindings[m].1, &hashes);
                    group_text.push_str(&format!("{}={};", names[m], text));
                }
                let group_hash = sha256_hex(group_text.as_bytes());
                for &m in &members {
                    hashes[m] = Some(sha256_hex(
                        format!("{}:{}", group_hash, names[m]).as_bytes(),
                    ));
                }
            }
            progress = true;
        }
    }

    let mut out = Vec::new();
    for i in 0..n {
        match &hashes[i] {
            Some(h) => out.push((names[i].clone(), h.clone())),
            None => return Err(format!("incremental: unhashed definition {}", names[i])),
        }
    }
    Ok(out)
}

/// Replace references to already-hashed siblings with hash markers (for
/// hashing only; the result is never evaluated or re-parsed).
/// Demand-driven change propagation (salsa/adapton early-cutoff, proposal
/// 0007): the set of definition names whose dependency-substituted content
/// hash DIFFERS between two program versions. By construction of the
/// dependency-substituted hash, this is exactly the edited definition plus its
/// transitive DEPENDENTS — independent siblings keep identical hashes (early
/// cutoff: same hash => no recompute). Both versions must declare the same
/// definition names.
pub fn changed_between(source_a: &str, source_b: &str) -> Result<Vec<String>, String> {
    let ha = definition_hashes(source_a)?;
    let hb = definition_hashes(source_b)?;
    let mut changed = Vec::new();
    for (name, hash_a) in &ha {
        let hash_b = hb
            .iter()
            .find(|(n, _)| n == name)
            .map(|(_, h)| h.clone());
        match hash_b {
            Some(hb2) if &hb2 == hash_a => {}
            _ => changed.push(name.clone()),
        }
    }
    changed.sort();
    Ok(changed)
}

fn substitute_sibling_refs(
    e: &px::PxExpr,
    names: &[String],
    hashes: &Vec<Option<String>>,
) -> px::PxExpr {
    match e {
        px::PxExpr::DeferredError(_) => e.clone(),
        px::PxExpr::Var(name) => {
            for (j, sib) in names.iter().enumerate() {
                if sib == name {
                    if let Some(h) = &hashes[j] {
                        return px::PxExpr::Var(format!("#{}", h));
                    }
                }
            }
            e.clone()
        }
        px::PxExpr::Int(_) | px::PxExpr::Float(_) | px::PxExpr::Bool(_) | px::PxExpr::Null => e.clone(),
        px::PxExpr::With { .. } => e.clone(),
        px::PxExpr::Str(parts) => px::PxExpr::Str(
            parts
                .iter()
                .map(|p| match p {
                    px::PxStrPart::Lit(s) => px::PxStrPart::Lit(s.clone()),
                    px::PxStrPart::Sub(sub) => {
                        px::PxStrPart::Sub(substitute_sibling_refs(sub, names, hashes))
                    }
                })
                .collect(),
        ),
        px::PxExpr::List(items) => px::PxExpr::List(
            items
                .iter()
                .map(|i| substitute_sibling_refs(i, names, hashes))
                .collect(),
        ),
        px::PxExpr::Select { base, name } => px::PxExpr::Select {
            base: Box::new(substitute_sibling_refs(base, names, hashes)),
            name: name.clone(),
        },
        px::PxExpr::Lambda { param, body } => px::PxExpr::Lambda {
            param: param.clone(),
            body: std::rc::Rc::new(substitute_sibling_refs(body, names, hashes)),
        },
        px::PxExpr::Apply { func, arg } => px::PxExpr::Apply {
            func: Box::new(substitute_sibling_refs(func, names, hashes)),
            arg: Box::new(substitute_sibling_refs(arg, names, hashes)),
        },
        px::PxExpr::If { cond, then_e, else_e } => px::PxExpr::If {
            cond: Box::new(substitute_sibling_refs(cond, names, hashes)),
            then_e: Box::new(substitute_sibling_refs(then_e, names, hashes)),
            else_e: Box::new(substitute_sibling_refs(else_e, names, hashes)),
        },
        px::PxExpr::Binary { op, lhs, rhs } => px::PxExpr::Binary {
            op: op.clone(),
            lhs: Box::new(substitute_sibling_refs(lhs, names, hashes)),
            rhs: Box::new(substitute_sibling_refs(rhs, names, hashes)),
        },
        px::PxExpr::LetIn { bindings, body } => {
            // Inner let names shadow siblings; v0 keeps substitution simple by
            // not descending past a shadowing binder for shadowed names.
            let shadowed: Vec<String> = bindings.iter().map(|(n, _)| n.clone()).collect();
            let filtered: Vec<String> = names
                .iter()
                .filter(|n| !shadowed.iter().any(|s| s == *n))
                .cloned()
                .collect();
            let filtered_hashes: Vec<Option<String>> = names
                .iter()
                .zip(hashes.iter())
                .filter(|(n, _)| !shadowed.iter().any(|s| s == *n))
                .map(|(_, h)| h.clone())
                .collect();
            px::PxExpr::LetIn {
                bindings: bindings
                    .iter()
                    .map(|(n, v)| {
                        (n.clone(), substitute_sibling_refs(v, &filtered, &filtered_hashes))
                    })
                    .collect(),
                body: Box::new(substitute_sibling_refs(body, &filtered, &filtered_hashes)),
            }
        }
        px::PxExpr::Attrs(fields) => px::PxExpr::Attrs(
            fields
                .iter()
                .map(|(n, v)| (n.clone(), substitute_sibling_refs(v, names, hashes)))
                .collect(),
        ),
    }
}

/// Realisation-backed evaluation with early cutoff.
/// Returns (value_sha256, cutoff).
pub fn incremental_eval(
    source: &str,
    store_path: &str,
    granted: &[String],
) -> Result<(String, bool), String> {
    let record = ir::ir_of(source)?;
    let ir_hash = record.ir_sha256;

    let existing = interop::host_read_file(store_path, granted).unwrap_or_default();
    for line in existing.split('\n') {
        let cols: Vec<&str> = line.split('\t').collect();
        if cols.len() == 3 && cols[0] == ir_hash {
            return Ok((String::from(cols[1]), true));
        }
    }

    let value = px::px_run(source)?;
    let value_sha = sha256_hex(value.as_bytes());
    let witness = gate::eval_witness(source, granted)?;
    let mut updated = existing;
    if !updated.is_empty() && !updated.ends_with('\n') {
        updated.push('\n');
    }
    updated.push_str(&format!("{}\t{}\t{}\n", ir_hash, value_sha, witness.out_hash));
    interop::host_write_file(store_path, &updated, granted)?;
    Ok((value_sha, false))
}

/// Verifying cache (proposal, maps pnix-hy 30): on a cache HIT, RE-DERIVE the
/// value and confirm it matches the stored value hash. A plain cache trusts the
/// store (early cutoff); a verifying cache AUDITS it — catching a corrupted or
/// stale entry. Returns (value_hash, verified_hit): verified_hit=true means the
/// cached value was re-checked and agreed; an error means the store lies.
pub fn incremental_eval_verify(
    source: &str,
    store_path: &str,
    granted: &[String],
) -> Result<(String, bool), String> {
    let record = ir::ir_of(source)?;
    let ir_hash = record.ir_sha256;
    let value = px::px_run(source)?;
    let value_sha = sha256_hex(value.as_bytes());

    let existing = interop::host_read_file(store_path, granted).unwrap_or_default();
    for line in existing.split('\n') {
        let cols: Vec<&str> = line.split('\t').collect();
        if cols.len() == 3 && cols[0] == ir_hash {
            if cols[1] == value_sha {
                return Ok((value_sha, true));
            } else {
                return Err(format!(
                    "verifying cache: store entry for {} is {} but re-derived {}",
                    &ir_hash[0..12],
                    cols[1],
                    value_sha
                ));
            }
        }
    }
    Err(format!("verifying cache: no entry for {}", &ir_hash[0..12]))
}
